core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/promisehandlers/PortCompletedHandler.scala

```diff
@@ -29,27 +29,33 @@ trait PortCompletedHandler {
         msg.portId,
         input = msg.input
       )
-
-      val operatorExecution =
-        cp.executionState.getOperatorExecution(VirtualIdentityUtils.getPhysicalOpId(sender))
-      val workerExecution = operatorExecution.getWorkerExecution(sender)
-
-      // set the port on this worker to be completed
-      (if (msg.input) workerExecution.getInputPortExecution(msg.portId)
-       else workerExecution.getOutputPortExecution(msg.portId)).setCompleted()
-
-      // check if the port on this operator is completed
-      val isPortCompleted =
-        if (msg.input) operatorExecution.isInputPortCompleted(msg.portId)
-        else operatorExecution.isOutputPortCompleted(msg.portId)
-
-      if (isPortCompleted) {
-        cp.workflowExecutionController.markRegionCompletion(globalPortId)
-        cp.workflowExecutionController.executeNextRegions()
-      } else {
-        // if the port is not completed yet, do nothing
-        Future(())
+      cp.workflow.regionPlan.getRegionOfPortId(globalPortId) match {
+        case Some(region) =>
+          val regionExecution = cp.workflowExecution.getRegionExecution(region.id)
+          val operatorExecution =
+            regionExecution.getOperatorExecution(VirtualIdentityUtils.getPhysicalOpId(sender))
+          val workerExecution = operatorExecution.getWorkerExecution(sender)
+
+          // set the port on this worker to be completed
+          (if (msg.input) workerExecution.getInputPortExecution(msg.portId)
+           else workerExecution.getOutputPortExecution(msg.portId)).setCompleted()
+
+          // check if the port on this operator is completed
+          val isPortCompleted =
+            if (msg.input) operatorExecution.isInputPortCompleted(msg.portId)
+            else operatorExecution.isOutputPortCompleted(msg.portId)
+
+          if (isPortCompleted) {
+            //        cp.workflowExecutionController.markRegionCompletion(globalPortId)
+            logger.info("port is completed, tying to execute next regions")
+            cp.workflowExecutionController.executeNextRegions()
+          } else {
+            // if the port is not completed yet, do nothing
+            Future(())
+          }
+        case None => // do nothing.
```

Comment:
is this one an invalid case? If so, add a warning log?

Reply:
no, currently "start" and "end" ports are not part of a region. added a comment to clarify it.